server/src/main/java/org/apache/druid/segment/realtime/appenderator/BaseAppenderatorDriver.java

```diff
@@ -554,39 +554,49 @@ protected AppenderatorDriverAddResult append(
-                      + " Removing them from deep storage and checking if someone else beat us to publishing."
-                  );
-                } else {
-                  log.warn(
-                      "Transaction failure while publishing segments because of [%s]. Please check the overlord log."
-                      + " Removing them from deep storage and checking if someone else beat us to publishing.",
-                      publishResult.getErrorMsg()
-                  );
-                }
-
-                segmentsAndMetadata.getSegments().forEach(dataSegmentKiller::killQuietly);
+                // Publishing didn't affirmatively succeed. However, segments with our identifiers may still be active
+                // now after all, for two possible reasons:
+                //
+                // 1) A replica may have beat us to publishing these segments. In this case we want to delete the
+                //    segments we pushed (if they had unique paths) to avoid wasting space on deep storage.
+                // 2) We may have actually succeeded, but not realized it due to missing the confirmation response
+                //    from the overlord. In this case we do not want to delete the segments we pushed, since they are
+                //    now live!
 
                 final Set<SegmentIdWithShardSpec> segmentsIdentifiers = segmentsAndMetadata
                     .getSegments()
                     .stream()
                     .map(SegmentIdWithShardSpec::fromDataSegment)
                     .collect(Collectors.toSet());
 
-                if (usedSegmentChecker.findUsedSegments(segmentsIdentifiers)
-                                      .equals(Sets.newHashSet(segmentsAndMetadata.getSegments()))) {
-                  log.info("Our segments really do exist, awaiting handoff.");
+                final Set<DataSegment> activeSegments = usedSegmentChecker.findUsedSegments(segmentsIdentifiers);
+
+                if (activeSegments.equals(ourSegments)) {
+                  log.info("Could not publish segments, but checked and found them already published. Continuing.");
+
+                  // Clean up pushed segments if they are physically disjoint from the published ones (this means
+                  // they were probably pushed by a replica, and with the unique paths option).
+                  final boolean physicallyDisjoint = Sets.intersection(
+                      activeSegments.stream().map(DataSegment::getLoadSpec).collect(Collectors.toSet()),
+                      ourSegments.stream().map(DataSegment::getLoadSpec).collect(Collectors.toSet())
+                  ).isEmpty();
+
+                  if (physicallyDisjoint) {
+                    segmentsAndMetadata.getSegments().forEach(dataSegmentKiller::killQuietly);
+                  }
                 } else {
+                  // Our segments aren't active. Publish failed for some reason. Clean them up and then throw an error.
```

Comment:
Is it still worth to print the error message in `publishResult` as https://github.com/apache/incubator-druid/pull/7558/files#diff-f18d4b23ee86a1255737ab0b2da247b5L574?

Reply:
I don't think so, since it's part of the exception, which should get logged elsewhere.

Reply:
I think the full stack trace will be in the overlord log, but printing the error message here is for easier debugging. I think it’s still worth because sometimes it’s not easy to find the correct corresponding stack trace with just the task ID in the overlord log. What do you think?

Reply:
That change was added in https://github.com/apache/incubator-druid/pull/7035, and it doesn't print any stack trace, but instead it prints some locks are revoked in overlord log. If it just says "check the overlord log" when a task fails because of revoked locks, I think this wouldn't be easy to debug for users who are not familiar with enough with druid.

Reply:
I removed the logging because it seemed like there were two cases:

1. We were going to hit the "Our segments really do exist, awaiting handoff" case, in which case there is no need for a scary warning message (everything is fine).
2. We were going to hit the "Failed to publish segments because of [%s]." case, in which case the error message would get logged anyway since it's part of an exception. (Presumably some higher level code would log this exception. I didn't check this but I assumed it.)

Reply:
What do you think about the above analysis @jihoonson?